### PR TITLE
264 Passing incorrect password produces erroneous error message

### LINKED ISSUE #188: Dell M620 (VRTX) fails to get inventory with IDRAC build 2.63.60.62 (01)
Running simple playbook for testing fails on IDRAC 2.63.60.62 (01) on Dell M620 which is supposed to work as per readme

---
- hosts: localhost
  connection: local
  name: Get system inventory
  gather_facts: False

  tasks:
  - name: Get system inventory
    dellemc.openmanage.dellemc_get_system_inventory:
       idrac_ip:   "{{ idrac_ip }}"
       idrac_user: "{{ idrac_user }}"
       idrac_password:  "{{ idrac_pwd }}"

Results should be that it works correctly, however i get the following

[root@Elysium baremetal-provision]# ansible-playbook get_inventory.yaml -e @galaga.yml
[WARNING]: Unable to parse /root/networking-workshop/lab_inventory/hosts as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Get system inventory] *********************************************************************************************************************************************************

TASK [Get system inventory] *********************************************************************************************************************************************************
[WARNING]: The value 443 (type int) in a string field was converted to '443' (type string). If this does not look like what you expect, quote the entire value to ensure it does not
change.
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Could not find device driver for iDRAC with IP Address: 10.192.1.44"}

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

[root@Elysium baremetal-provision]# ansible-playbook get_inventory.yaml -e @galaga.yml -vvv
ansible-playbook 2.9.15
  config file = /root/.ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.6.8 (default, Nov 16 2020, 16:55:22) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
Using /root/.ansible.cfg as config file
host_list declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
auto declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /root/networking-workshop/lab_inventory/hosts as it did not pass its verify_file() method
[WARNING]: Unable to parse /root/networking-workshop/lab_inventory/hosts as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAYBOOK: get_inventory.yaml ********************************************************************************************************************************************************
1 plays in get_inventory.yaml

PLAY [Get system inventory] *********************************************************************************************************************************************************
META: ran handlers

TASK [Get system inventory] *********************************************************************************************************************************************************
task path: /git/baremetal-provision/get_inventory.yaml:8
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326 `" && echo ansible-tmp-1608703256.6670403-17492-271507268445326="` echo /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326 `" ) && sleep 0'
Using module file /root/.ansible/collections/ansible_collections/dellemc/openmanage/plugins/modules/dellemc_get_system_inventory.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-1748644kze0dq/tmp1m8gnpjq TO /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326/AnsiballZ_dellemc_get_system_inventory.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326/ /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326/AnsiballZ_dellemc_get_system_inventory.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326/AnsiballZ_dellemc_get_system_inventory.py && sleep 0'
eeeeeeeeeeeeeeeeee<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1608703256.6670403-17492-271507268445326/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_dellemc.openmanage.dellemc_get_system_inventory_payload_ojvkw230/ansible_dellemc.openmanage.dellemc_get_system_inventory_payload.zip/ansible_collections/dellemc/openmanage/plugins/modules/dellemc_get_system_inventory.py", line 90, in main
  File "/tmp/ansible_dellemc.openmanage.dellemc_get_system_inventory_payload_ojvkw230/ansible_dellemc.openmanage.dellemc_get_system_inventory_payload.zip/ansible_collections/dellemc/openmanage/plugins/module_utils/dellemc_idrac.py", line 49, in __enter__
    raise RuntimeError(msg)
[WARNING]: The value 443 (type int) in a string field was converted to '443' (type string). If this does not look like what you expect, quote the entire value to ensure it does not
change.
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "idrac_ip": "10.192.1.44",
            "idrac_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "idrac_port": "443",
            "idrac_user": "root"
        }
    },
    "msg": "Could not find device driver for iDRAC with IP Address: 10.192.1.44"

**Alternate Test**: 
When I test with another M620 blade with latest 2.65.65.65 (15) works as expected.

This is more of an informative issue ticket that not all builds of iDRAC work correctly i will upgrade to 2.65.65.65 on this server but figured it would be good for notes.

### LINKED ISSUE #264: Passing incorrect password produces erroneous error message
I passed the following input to the `idrac_network.py` module:

```
{
    "ANSIBLE_MODULE_ARGS": {
        "idrac_ip": "192.168.1.63",
        "idrac_user": "root",
        "idrac_password": "calvin",
        "share_name": "some_share",
        "share_user": "some_username",
        "share_password": "some_password"
    }
}
```

This produces an erroneous error message that is misleading to the user:

`{"failed": true, "msg": "Could not find device driver for iDRAC with IP Address: 192.168.1.63", "invocation": {"module_args": {"idrac_ip": "192.168.1.63", "idrac_user": "root", "idrac_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "share_name": "some_share", "share_user": "some_username", "share_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "idrac_port": 443, "share_mnt": null, "register_idrac_on_dns": null, "dns_idrac_name": null, "auto_config": null, "static_dns": null, "setup_idrac_nic_vlan": null, "vlan_id": null, "vlan_priority": null, "enable_nic": null, "nic_selection": null, "failover_network": null, "auto_detect": null, "auto_negotiation": null, "network_speed": null, "duplex_mode": null, "nic_mtu": null, "ip_address": null, "enable_dhcp": null, "enable_ipv4": null, "dns_from_dhcp": null, "static_dns_1": null, "static_dns_2": null, "static_gateway": null, "static_net_mask": null}}}`

If the user passes an incorrect password they should receive an error message which says as much instead of an error message related to device driver.

Providing the correct password clears the error.

### PR DESCRIPTION
**Pull request by grantcurell**
No one but a developer would know what "Could not find device driver for iDRAC with IP Address" means. This should not be an error message we display to a user. This PR replaces the error message with the two actual causes of this error message - either a bad username/password or the wrong IP address.

Fixes #188
Fixes #264

Signed-off-by: Grant Curell <grant_curell@dell.com>

### CODE REVIEW

File: plugins/module_utils/dellemc_idrac.py [hunk 1]
```diff
@@ -45,7 +45,11 @@ def __enter__(self):
         self.sdk.importPath()
         self.handle = self.sdk.get_driver(self.sdk.driver_enum.iDRAC, self.idrac_ip, self.creds, pOptions=self.pOp)
         if self.handle is None:
-            msg = "Could not find device driver for iDRAC with IP Address: {0}".format(self.idrac_ip)
+            msg = "Unable to create a driver to communicate with idrac %s. This most often indicates one of the " \
```
Comment (rajeevarakkal) [inline plugins/module_utils/dellemc_idrac.py:48]: @grantcurell - reword this to Unable to communicate with idrac %s. This typically indicates a bad " \

Basically you even remove the create a driver words from your suggestion.

**Review by jagadeeshnv — CHANGES_REQUESTED** [review]
@grantcurell 

1. Please replace idrac with iDRAC
2. Remove "(maybe due....)" 
3. "Unable to create a driver for iDRAC with IP %s"

We can accommodate this PR soon.

File: plugins/module_utils/dellemc_idrac.py [hunk 2]
```diff
@@ -45,7 +45,11 @@ def __enter__(self):
...
+                  "following:\n\n" \
+                  '1. A bad username/password\n' \
+                  '2. An incorrect IP address\n' \
+                  '3. A failure in a TLS/SSL handshake (maybe due to a proxy/firewall?)' % self.idrac_ip
```
Comment (rajeevarakkal) [inline plugins/module_utils/dellemc_idrac.py:52]: @grantcurell  - instead of TLS/SSL handshake - reword probably to `unreachable iDRAC  <idrac_ip> `

Comment (grantcurell) [reply plugins/module_utils/dellemc_idrac.py:52]: Unreachable IP address is accounted for #2. I was asked to add the TLS/SSL handshake issue specifically because it has come up frequently for support

File: plugins/module_utils/dellemc_idrac.py (hunk 1 above)
Comment (grantcurell) [reply plugins/module_utils/dellemc_idrac.py]: Finished

File: plugins/module_utils/dellemc_idrac.py [hunk 3]
```diff
@@ -45,7 +45,8 @@ def __enter__(self):
         self.sdk.importPath()
         self.handle = self.sdk.get_driver(self.sdk.driver_enum.iDRAC, self.idrac_ip, self.creds, pOptions=self.pOp)
         if self.handle is None:
-            msg = "Could not find device driver for iDRAC with IP Address: {0}".format(self.idrac_ip)
+            msg = "Unable to create a driver to communicate with idrac %s. This typically indicates a bad " \
```
[the remark this answers is not in the document]
Comment (grantcurell) [reply plugins/module_utils/dellemc_idrac.py]: Finished

File: plugins/module_utils/dellemc_idrac.py (hunk 2 above)
Comment (grantcurell) [reply plugins/module_utils/dellemc_idrac.py]: I updated it to:

```python
            msg = "Unable to communicate with iDRAC %s. This most often indicates one of the " \
                  "following:\n\n" \
                  '1. A bad username/password\n' \
                  '2. The iDRAC IP address is unreachable\n' \
                  '3. A failure in a TLS/SSL handshake (proxies/firewalls are a common cause of this problem)' % self.idrac_ip                  ```

**Review by jagadeeshnv — APPROVED** [review]

File: plugins/module_utils/dellemc_idrac.py (hunk 2 above)
Comment (jagadeeshnv) [reply plugins/module_utils/dellemc_idrac.py]: Thank you, The content is fine, we will get it reviewed by our Infodev team and get back on it

**Review by jagadeeshnv — COMMENTED** [review]
Thank you, The content is fine, we will get it reviewed by our Infodev team and get back on it

**Review by jagadeeshnv — CHANGES_REQUESTED** [review]
@grantcurell 
**"Unable to communicate with iDRAC {0}. This may be due to one of the following: Incorrect username or password, unreachable iDRAC IP or a failure in TLS/SSL handshake."** 
This is the modified content, please update to this

**Comment by jagadeeshnv** [comment]
@grantcurell Please update as in the previous comment. We removed newline character, as its rarely used in any ansible modules. Let us know if specific use case with such a scenario

**Comment by grantcurell** [comment]
> @grantcurell Please update as in the previous comment. We removed newline character, as its rarely used in any ansible modules. Let us know if specific use case with such a scenario

Support specifically told me proxies and firewalls are usually the cause for the third error. Is there a reason we want to reduce the amount of information we provide to the user if we have a decent guess what the problem is?

**Comment by jagadeeshnv** [comment]
@grantcurell handshake failure can be due to multiple reasons.

**Comment by grantcurell** [comment]
Since the user hint on the most likely cause seems to be a major issue, I removed it.

**Review by jagadeeshnv — APPROVED** [review]

**Review by rajeevarakkal — APPROVED** [review]
